output/cloud/expv2/mapping.go

```diff
@@ -0,0 +1,120 @@
+package expv2
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mstoykov/atlas"
+	"go.k6.io/k6/metrics"
+	"go.k6.io/k6/output/cloud/expv2/pbcloud"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// TODO: unit test
+func mapTimeSeriesLabelsProto(timeSeries metrics.TimeSeries, testRunID string) []*pbcloud.Label {
+	labels := make([]*pbcloud.Label, 0, ((*atlas.Node)(timeSeries.Tags)).Len()+2)
+	labels = append(labels,
+		&pbcloud.Label{Name: "__name__", Value: timeSeries.Metric.Name},
+		&pbcloud.Label{Name: "test_run_id", Value: testRunID})
+
+	// TODO: move it as a shared func
+	// https://github.com/grafana/k6/issues/2764
+	n := (*atlas.Node)(timeSeries.Tags)
+	if n.Len() < 1 {
+		return labels
+	}
+	for !n.IsRoot() {
+		prev, key, value := n.Data()
+		labels = append(labels, &pbcloud.Label{Name: key, Value: value})
+		n = prev
+	}
+	return labels
+}
+
+// TODO: unit test
+func mapMetricTypeProto(mt metrics.MetricType) pbcloud.MetricType {
+	var mtype pbcloud.MetricType
+	switch mt {
+	case metrics.Counter:
+		mtype = pbcloud.MetricType_METRIC_TYPE_COUNTER
+	case metrics.Gauge:
+		mtype = pbcloud.MetricType_METRIC_TYPE_GAUGE
+	case metrics.Rate:
+		mtype = pbcloud.MetricType_METRIC_TYPE_RATE
+	case metrics.Trend:
+		mtype = pbcloud.MetricType_METRIC_TYPE_TREND
+	}
+	return mtype
+}
+
+// TODO: unit test
+func addBucketToTimeSeriesProto(
+	timeSeries *pbcloud.TimeSeries,
+	mt metrics.MetricType,
+	time time.Time,
+	sink metrics.Sink,
+) {
+	if timeSeries.Samples == nil {
+		initTimeSeriesSamples(timeSeries, mt)
+	}
+
+	switch mt {
+	case metrics.Counter:
+		c := sink.(*metrics.CounterSink).Value //nolint: forcetypeassert
```

Comment:
I wonder if it is not going to be as fast to just do the type select and drop the assert and the nolint?

Have you tried it?

Reply:
https://github.com/grafana/k6/pull/3085/files#r1209311635 moved here